File: CLAUDE/scam-slingshot/_tools/scenarios/p1-r6-bandvfx.mjs
/**
 * p1-r6-bandvfx.mjs — THE r5-critic METRIC, made measurable.
 *
 * "During the first 150 ms the recoiling BAND — not a smoke cloud — must be the
 *  highest-contrast thing at the sling. Band pixels must outnumber launch-VFX pixels in a
 *  520 px box on the pouch."
 *
 * Method (honest isolation, all three renders of the SAME simulated instant via SS.__render):
 *   A        = the frame as shipped
 *   A_noFX   = same instant, every fx-* pool hidden      -> diff(A,A_noFX) = VFX pixels ACTUALLY
 *                                                           VISIBLE in the composite
 *   A_noBand = same instant, both band straps hidden     -> diff(A,A_noBand) = BAND pixels ACTUALLY
 *                                                           VISIBLE in the composite
 * Both diffs are taken against the SAME shipped frame, so a band buried under an opaque plume
 * correctly counts as zero band pixels and the ratio cannot be gamed by drawing the band on top
 * of something that hides it.
 *
 * Writes geom.json (pouch px, AD px, device pixel ratio) + A/NOFX/NOBAND PNGs per timestamp.
 * Analysed by _tools/p1-r6-bandvfx.py.
 */
export default async ({ page, game, OUT }) => {
  const say = (k, v) => console.log('### ' + k + ' ' + JSON.stringify(v));
  const TS = [0, 30, 60, 100, 150];

  // Default shot is the one every r6 measurement was taken on. P1_ANGLE / P1_POWER override it
  // so the same metric can be swept across the draw without forking the scenario — a fan that
  // only wins the pixel ratio at one draw has not won it.
  const ANGLE = +(process.env.P1_ANGLE ?? 0.42);
  const POWER = +(process.env.P1_POWER ?? 0.90);
  say('shot', { angle: ANGLE, power: POWER });
  await game('SS.seed(11); await SS.seek(2000);');
  await game('await SS.aim({angle:args[0], power:args[1]}); await SS.seek(400);', ANGLE, POWER);

  const info = await game('return await SS.release();');
  say('release', { ad: info.ad, muzzleAD: info.muzzleAD, power: info.power });

  /**
   * The box has to be re-centred at EVERY timestamp, not pinned to the release-time pixel.
   * The camera pans and dollies during flight (P4), so a fixed box slides off the sling: on
   * the 0.60 rad / 0.60 draw sweep it had left the band entirely by t = 150 ms and reported
   * `bandpx 0`, i.e. "the band is invisible", for a band that was ringing in plain sight
   * 300 px away. Same class of error as the stale drag pixels in capture.mjs's aimPx note.
   */
  const probe = () => game(`
    const w = SS.__world, s = w.sling, cam = w.camera;
    const r = w.renderer.domElement.getBoundingClientRect();
    const V3 = cam.position.constructor;
    const P = (x,y) => { const v = new V3(x,y,0).project(cam);
      return [ (v.x*0.5+0.5)*r.width, (-v.y*0.5+0.5)*r.height ]; };
    const halfH = Math.abs(cam.position.z) * Math.tan(cam.fov*Math.PI/360);
    return { anchorPx: P(s.anchor.x, s.anchor.y),
             pxPerWorld: (r.height/2)/halfH,
             cssW: r.width, cssH: r.height, dpr: devicePixelRatio };`);
  const geom = await probe();
  const adPx = info.ad * geom.pxPerWorld;
  say('geom', { ...geom, adPx: +adPx.toFixed(2) });
  const perT = {};

  // Visibility toggles. FX pools are scene children named fx-*; the two band straps hang off
  // the sling. Nothing is stepped between the three renders, so all three are the same instant.
  const setVis = (what, on) => game(`
    const w = SS.__world;
    if (args[0] === 'fx') {
      w.scene.traverse(o => { if (o.name && o.name.startsWith('fx-')) o.visible = args[1]; });
    } else {
      for (const b of w.sling.bands) b.tube.group.visible = args[1];
    }
    SS.__render();
    return true;`, what, on);

  let n = 0;
  const cap = async (name) => {
    await page.screenshot({ path: `${OUT}/${name}.png` });
    n++;
  };

  let t = 0;
  for (const want of TS) {
    if (want > t) { await game('await SS.seek(args[0]);', want - t); t = want; }
    await game('SS.__render();');
    const p = await probe();
    perT[want] = { anchorPx: p.anchorPx, adPx: info.ad * p.pxPerWorld };
    await cap(`A-t${want}`);
    await setVis('fx', false);  await cap(`NOFX-t${want}`);  await setVis('fx', true);
    await setVis('band', false); await cap(`NOBAND-t${want}`); await setVis('band', true);
    await game('SS.__render();');
  }

  const fs = await import('node:fs/promises');
  await fs.writeFile(`${OUT}/geom.json`, JSON.stringify(
    { ...geom, adPx, ts: TS, ad: info.ad, muzzleAD: info.muzzleAD,
      angle: ANGLE, power: POWER, perT }, null, 2));
  say('shots', n);
  console.log('### DONE');
};

File: CLAUDE/scam-slingshot/_tools/p1-r6-bandvfx.py
#!/usr/bin/env python3
"""
p1-r6-bandvfx.py — scores the r5 critic's gap metric from the PNGs p1-r6-bandvfx.mjs writes.

For each timestamp it counts, inside a square box centred on the sling anchor:
  bandpx = pixels where hiding the two straps changed the SHIPPED frame
  vfxpx  = pixels where hiding every fx-* pool changed the SHIPPED frame
and reports vfx:band. The critic measured 30:1 at t=0 and 103:1 at t=100. Target: BAND WINS
(ratio < 1) through the first 150 ms.

Also reports where the VFX actually lives — its 90th-percentile reach along the shot axis in
AD — because "confined to roughly 2 AD of the pouch" is the shape half of the ask, and a fan
that hits the pixel ratio by going FAINT rather than by going SMALL has not done the job.

Usage: python3 p1-r6-bandvfx.py <shots-dir> [box-css-px ...]
"""
import json, os, sys
import numpy as np
from PIL import Image

D = sys.argv[1]
BOXES = [float(x) for x in sys.argv[2:]] or [520.0]
g = json.load(open(os.path.join(D, 'geom.json')))
ax_css, ay_css = g['anchorPx']
adPx_css = g['adPx']

def arr(p):
    return np.asarray(Image.open(p).convert('RGB'), dtype=np.int16)

first = arr(os.path.join(D, f"A-t{g['ts'][0]}.png"))
H, W = first.shape[:2]
S = W / g['cssW']                      # device px per css px (deviceScaleFactor)
ax, ay, adPx = ax_css * S, ay_css * S, adPx_css * S
ys, xs = np.mgrid[0:H, 0:W]
print(f"frame {W}x{H}  scale {S:g}x css   anchor=({ax:.0f},{ay:.0f}) devpx   "
      f"AD={adPx:.1f} devpx   muzzle={g['muzzleAD']} AD")
print()

perT = g.get('perT', {})

def centre(t):
    """Box centre + AD for this timestamp. The camera pans during flight, so a box pinned to
    the release-time pixel walks off the sling and starts reporting the band as invisible."""
    e = perT.get(str(t)) or perT.get(t)
    if not e:
        return ax, ay, adPx
    return e['anchorPx'][0]*S, e['anchorPx'][1]*S, e['adPx']*S

for box in BOXES:
    half = box / 2.0
    inbox = (np.abs(xs - ax) <= half) & (np.abs(ys - ay) <= half)
    print(f"--- box {box:.0f} px  (side = {box/adPx:.2f} AD)  {int(inbox.sum())} px ---")
    print("   t    bandpx     vfxpx    vfx:band    verdict     vfx p90 reach (AD from anchor)")
    for t in g['ts']:
        cx, cy, cad = centre(t)
        inbox = (np.abs(xs - cx) <= half) & (np.abs(ys - cy) <= half)
        A = arr(os.path.join(D, f'A-t{t}.png'))
        nf = arr(os.path.join(D, f'NOFX-t{t}.png'))
        nb = arr(os.path.join(D, f'NOBAND-t{t}.png'))
        dv = (np.abs(A - nf).max(axis=2) > 8)
        db = (np.abs(A - nb).max(axis=2) > 8)
        bp, vp = int((db & inbox).sum()), int((dv & inbox).sum())
        ratio = (vp / bp) if bp else float('inf')
        ok = 'BAND WINS' if ratio < 1 else f'vfx {ratio:5.1f}:1'
        # where the VFX lives, measured over the WHOLE frame, in AD from the anchor
        yy, xx = np.nonzero(dv)
        reach = (np.percentile(np.hypot(xx - cx, yy - cy), 90) / cad) if len(xx) else 0.0
        print(f"{t:4d}  {bp:8d}  {vp:8d}   {ratio:8.2f}   {ok:>12s}      {reach:6.2f}")
    print()
